src/workspace/mod.rs

```diff
@@ -4391,24 +4401,137 @@ mod test {
   }
 
   #[test]
-  fn workspace_discovery_deno_json_cache() {
-    #[derive(Default)]
-    struct Cache(RefCell<HashMap<PathBuf, ConfigFileRc>>);
-
-    impl DenoJsonCache for Cache {
-      fn get(&self, path: &Path) -> Option<ConfigFileRc> {
-        self.0.borrow().get(path).cloned()
+  fn test_config_not_deno_workspace_member() {
+    fn assert_err(err: &WorkspaceDiscoverError, config_file_path: &Path) {
+      match err.as_kind() {
+        WorkspaceDiscoverErrorKind::ConfigNotWorkspaceMember {
+          workspace_url,
+          config_url,
+        } => {
+          assert_eq!(
+            workspace_url.as_ref(),
+            &Url::from_directory_path(root_dir()).unwrap()
+          );
+          assert_eq!(
+            config_url,
+            &Url::from_file_path(config_file_path).unwrap()
+          );
+        }
+        _ => unreachable!(),
       }
+    }
 
-      fn set(&self, path: PathBuf, deno_json: ConfigFileRc) {
-        self.0.borrow_mut().insert(path, deno_json);
-      }
+    for file_name in ["deno.json", "deno.jsonc"] {
+      let config_file_path = root_dir().join("member-b").join(file_name);
+      let mut fs = TestFileSystem::default();
+      fs.insert_json(
+        root_dir().join("deno.json"),
+        json!({
+          "workspace": ["./member-a"],
+        }),
+      );
+      fs.insert_json(root_dir().join("member-a/deno.json"), json!({}));
+      fs.insert_json(config_file_path.clone(), json!({}));
+      let err = workspace_at_start_dir_err(&fs, &root_dir().join("member-b"));
+      assert_err(&err, &config_file_path);
+
+      // try for when the config file is specified as well
+      let err = WorkspaceContext::discover(
+        WorkspaceDiscoverStart::ConfigFile(&config_file_path),
+        &WorkspaceDiscoverOptions {
+          fs: &fs,
+          discover_pkg_json: true,
+          ..Default::default()
+        },
+      )
+      .unwrap_err();
+      assert_err(&err, &config_file_path);
+    }
+  }
+
+  #[test]
+  fn test_config_not_deno_workspace_member_non_natural_config_file_name() {
```

Comment:
Test for bug fix: https://github.com/denoland/deno_config/pull/95/files#r1683560986